check if any certificate needs to be renewed

leipert opened the pull request:
@travismiller was faster on automating the pages workflow, thank you! 🎉 

On my local branch, I added a quick check, if certificate renewal is even needed at all.
High level workflow is like this

Only get a new certificate, if at least one of the pages given in the arguments:

1. is not registered yet
2. does not have a certificate
3. has an expired certificate
4. has an certificate expiring within X days (default 15)

```bash
$ ./index.js --domain xyz.leipert.io --domain leipert.io \
    --email xxx@yyy.io \
    --repository https://gitlab.com/leipert/leipert.gitlab.io \
    --path /public/.well-known/acme-challenge
All domains xyz.leipert.io, leipert.io have a valid certificate (expiration in more than 15 days)
```

leipert committed:
check if any certificate needs to be renewed

tmaier commented:
Great idea. Consider to increase the default to 31 days, as this fits better to the Pipeline Scheduling feature supported if #36 gets merged.

The updated README suggests to run the pipeline once per month. With a default of 15, the user would miss the renewal date.

leipert committed:
Change default expiration to 31 days from 15

rolodato commented:
Thanks for the PR! I would make the following changes for consistency with `certbot renew`:

* Default expiration should be 30 days
* No option to configure a default expiration
* Add a `--force-renewal` option that bypasses the expiration check and renews anyway

tmaier commented:
The reasons why I proposed 31 instead of 30 are
- some months have 31 days (with 30, they would not request a new certificate)
- you could come into a race condition where the previous build and validation step took longer than the current one (can be just a few minutes). In this case the the delta would be higher than 30 days and the renewal would be skipped

leipert commented:
I have to say, that I agree with @rolodato on that one:
1. Consistency with certbot is a good idea.
2. if you run `gitlab-le` with a cronjob each day, it doesn't matter whether it's 30 or 31 days.

leipert committed:
change minimum valid days to 30 days

leipert committed:
implement parameter `force-renewal`

leipert committed:
ensure that the script exist with `0` if all certificates are valid

leipert commented:
I updated the behavior as suggested. Additionally I made sure that the exit code is `0` if all certificates are valid

leipert commented:
/ping @rolodato

rolodato commented on args.js:
```diff
@@ -26,6 +26,10 @@ module.exports = yargs
         type: 'boolean',
         default: false
+    }).option('force-renewal', {
+        describe: 'Force renewal of certificate, even if it expires in more than 30 days',
```
This should be phrased with plural instead:

> Force renewal of certificates, even if they expire in more than 30 days.

leipert replied:
Two issues with that:
1. `--production` also is in singular:

    > Obtain a real certificate instead of a dummy one
    
2. Technically you only obtain one certificate per execution, right?

Should I still change the text?

rolodato commented on package.json:
```diff
@@ -9,6 +9,7 @@
   "dependencies": {
     "bluebird": "3.5.0",
     "le-acme-core": "https://git.daplie.com/rolodato/le-acme-core/repository/archive.tar.gz",
+    "moment": "^2.20.1",
```
I don't think it's necessary to introduce this dependency, we can do a little date arithmetic instead :)

leipert replied:
removed the dependency

rolodato commented on lib.js:
```diff
@@ -6,9 +6,12 @@ const ms = require('ms');
 const path = require('path');
 const { URL } = require('url');
 
+const EXPIRATION_IN_DAYS = 30;
```
Can be repliaced by `const EXPIRATION = ms('30 days')` if we get rid of the moment dependency.

leipert replied:
removed the dependency

rolodato commented on lib.js:
```diff
@@ -111,6 +133,16 @@ module.exports = (options) => {
+
+            if (needsNoRenewal) {
+                console.log(`All domains (${options.domain.join(', ')}) have a valid certificate (expiration in more than ${EXPIRATION_IN_DAYS} days)`);
+                process.exit(0);
```
I would try to avoid this `process.exit` here - this file should only be used as a library for the main program (main.js). I think the better approach would be to refactor this file so the exported function (`getCertificates`) resolves successfully but with an empty array as a result, which means that everything went OK but no new certificates needed to be generated.

rolodato commented on lib.js:
```diff
@@ -99,8 +102,27 @@ module.exports = (options) => {
     };
 
+    const hasValidCertificate = (pagesDomain) => {
+        if (options.domain.includes(pagesDomain.domain)) {
```
This function should not care about `option.domain`,  it should only care about the domain passed as a parameter. If we need to filter other domains it should be done at the call site instead of here.

rolodato commented on lib.js:
```diff
@@ -99,8 +102,27 @@ module.exports = (options) => {
 
+    const hasValidCertificate = (pagesDomain) => {
+        if (options.domain.includes(pagesDomain.domain)) {
+            if (pagesDomain.certificate) {
```
Can we avoid nesting these ifs so deeply? :)

leipert committed:
remove dependency to moment

leipert committed:
reduce complexity of `hasValidCertificate`

leipert committed:
refactor renewal check to not exit process, if renewal is not necessary

leipert commented:
@rolodato, I fixed all the issues you described in your review (except the singular/plural one). 
For the `process.exit` change, I had to refactor a bit:

I moved registering of ACME account, certificate retrieval and certificate upload into a new function `runACMEWorkflow`.

leipert committed:
use easier `return` instead of `then(() =>`

leipert commented:
Hey @rolodato, 

I thought, I'd ping again ;)

leipert commented:
@rolodato Ping ping :P
